sound/soc/sof/topology.c

```diff
@@ -1276,10 +1284,20 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 	swidget->pipeline_id = index;
 	swidget->private = NULL;
 
-	dev_dbg(scomp->dev, "tplg: ready widget id %d pipe %d type %d name : %s stream %s\n",
-		swidget->comp_id, index, swidget->id, tw->name,
-		strnlen(tw->sname, SNDRV_CTL_ELEM_ID_NAME_MAXLEN) > 0
-			? tw->sname : "none");
+	ret = sof_parse_tokens(scomp, swidget, comp_pin_tokens,
```

Comment:
I think we should initialize the num_source_pins/num_sink_pins to 1 by default and then override it with values from topology after we parse the tokens if they do exist.

Reply:
I think this can be done in topology, we have the default value of those two tokens in widget-common.conf, and a widget can override the default value with in the widget.conf, for example, copier can override the default value in copier.conf

Reply:
We should anyway make sure num_source_pins and num_sink_pins >= 1 before calling ida_alloc_max. Besides, kernel should make sure it will not be broken if user load an invalid topology.